### PR TITLE
[AppContext] Add register, invoke factory @open sesame 11/26 18:15

### PR DESCRIPTION
**Pull request by zhoonit**
- ~#728 [AppContext] Add AppContext~
- ~#720 [Fix] Throw when read/save fails~
- ~#721  [Optimizer] Change enum type to string~
- [AppContext] Add register, invoke factory
```
  [AppContext] Add registerer,invoke factory methods
    
    **Changes proposed in this PR:**
    - Add factory registerer
    - Add factory invoker
    - Register built-in objects to each layers(postponed)
    - Add tests
    
    **Self evaluation:**
    1. Build test: [X]Passed [ ]Failed [ ]Skipped
    2. Run test: [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>
```

### CODE REVIEW

File: nntrainer/app_context.h [hunk 1]
```diff
@@ -52,7 +87,72 @@ class AppContext {
...
+   * @param int_key key to access the factory by integer, if it is -1(default),
+   * the function automatically unsigned the key and return
+   * @return const int unique integer value to access the current factory
+   * @throw
+   */
+  template <typename T>
+  const int registerFactory(const std::string &key,
+                            const FactoryType<T> &factory,
+                            const int int_key = -1) {
+
```
Comment (zhoonit) [inline nntrainer/app_context.h:114]: Add mutex here

**Comment by jijoongmoon** [comment]
check the ci please?

File: test/unittest/unittest_nntrainer_appcontext.cpp [hunk 2]
```diff
@@ -83,6 +88,145 @@ TEST_F(nntrainerAppContextDirectory, notExisitingSetDirectory_n) {
...
+  EXPECT_THROW(ac.createObject<ml::train::Optimizer>("unknown"),
+               std::runtime_error);
+}
+
+TEST(nntrainerAppContextObjs, callingUnknownFactoryOptimizerWithIntKey_n) {
+  auto ac = nntrainer::AppContext();
+
+  int num = ac.registerFactory(
+    nntrainer::AppContext::unknownFactory<ml::train::Optimizer>, "unknown",
+    999);
```
Comment (kparichay) [inline test/unittest/unittest_nntrainer_appcontext.cpp:225]: add a check that `num == 999`?

File: nntrainer/app_context.h [hunk 3]
```diff
@@ -52,9 +94,110 @@ class AppContext {
...
+      ss << "Int Key is not found for the object. Key: " << int_key;
+      throw exception::not_supported(ss.str().c_str());
+    }
+
+    return createObject<T>(entry->second, props);
+  }
+
+  template <typename T>
+  PtrType<T> createObject(const std::string &key, const PropsType &props = {}) {
+    auto &index = std::get<IndexType<T>>(factory_map);
```
Comment (kparichay) [inline nntrainer/app_context.h:171]: sorry for the basic question - but I wonder if factory_map can contain multiple tuples? as factory_map is limited to just `optimizer`.

File: nntrainer/app_context.h [hunk 4]
```diff
@@ -52,9 +94,110 @@ class AppContext {
    */
   const std::string getWorkingPath(const std::string &path = "");
 
+  template <typename T>
+  const int registerFactory(const PtrFactoryType<T> factory,
+                            const std::string &key = "",
+                            const int int_key = -1) {
+    FactoryType<T> f = factory;
```
Comment (kparichay) [inline nntrainer/app_context.h:101]: also I wonder if this is auto resolved with compiler?

File: nntrainer/app_context.h [hunk 5]
```diff
@@ -52,9 +94,110 @@ class AppContext {
...
+    int_map[assigned_int_key] = assigned_key;
+
+    ml_logd("factory has registered with key: %s, int_key: %d",
+            assigned_key.c_str(), assigned_int_key);
+
+    return assigned_int_key;
+  }
+
+  template <typename T>
+  PtrType<T> createObject(const int int_key, const PropsType &props = {}) {
```
Comment (kparichay) [inline nntrainer/app_context.h:154]: for now, this is ok, but later you can consider `std::vector<std::string>` than `PropsType` for interface as former way is more readable IMO.

File: test/unittest/unittest_nntrainer_appcontext.cpp (hunk 2 above)
Comment (zhoonit) [reply test/unittest/unittest_nntrainer_appcontext.cpp:225]: Done

File: nntrainer/app_context.h (hunk 3 above)
Comment (zhoonit) [reply nntrainer/app_context.h:171]: Yes, you can check the same part #753 that layer is added :)

Comment (kparichay) [reply nntrainer/app_context.h:171]: got it.

File: test/unittest/unittest_nntrainer_appcontext.cpp (hunk 2 above)
Comment (zhoonit) [reply test/unittest/unittest_nntrainer_appcontext.cpp:226]: Done

File: nntrainer/app_context.h (hunk 5 above)
Comment (zhoonit) [reply nntrainer/app_context.h:154]: Done :)

**Review by kparichay — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**